PackageExplorer/Controls/MultiSelectTreeView.cs

```diff
@@ -0,0 +1,334 @@
+                {
+                    SelectMultipleItemsRandomly(treeView, treeViewItem);
+                }
+                else if (Keyboard.Modifiers == ModifierKeys.Shift)
+                {
+                    SelectMultipleItemsContinuously(treeView, treeViewItem);
+                }
+                else
+                {
+                    SelectSingleItem(treeView, treeViewItem);
+                }
+            }
+        }
+
+        private static void OnTreeViewItemPreviewMouseDown(object sender, MouseEventArgs e)
+        {
+            var treeViewItem = FindTreeViewItem(e.OriginalSource as DependencyObject);
+
+            if (treeViewItem != null && treeViewItem.IsFocused)
+                OnTreeViewItemGotFocus(sender, e);
+        }
+
+        private static void OnTreeViewItemPreviewMouseUp(object sender, MouseButtonEventArgs e)
+        {
+            var treeViewItem = FindTreeViewItem(e.OriginalSource as DependencyObject);
+
+            if (treeViewItem == _selectTreeViewItemOnMouseUp)
+            {
+                SelectItems(treeViewItem, sender as TreeView);
+            }
+        }
+
+        private static TreeViewItem? FindTreeViewItem(DependencyObject? dependencyObject)
+        {
+            if (!(dependencyObject is Visual || dependencyObject is Visual3D))
+                return null;
+
+            var treeViewItem = dependencyObject as TreeViewItem;
+            if (treeViewItem != null)
+            {
+                return treeViewItem;
+            }
+
+            return FindTreeViewItem(VisualTreeHelper.GetParent(dependencyObject));
+        }
+
+        private static void SelectSingleItem(TreeView treeView, TreeViewItem treeViewItem)
+        {
+            // first deselect all items
+            DeSelectAllItems(treeView, null);
+            SetIsItemSelected(treeViewItem, true);
+            SetStartItem(treeView, treeViewItem);
+        }
+
+        private static void DeSelectAllItems(TreeView? treeView, TreeViewItem? treeViewItem)
+        {
+            if (treeView != null)
+            {
+                for (int i = 0; i < treeView.Items.Count; i++)
+                {
+                    var item = treeView.ItemContainerGenerator.ContainerFromIndex(i) as TreeViewItem;
+                    if (item != null)
+                    {
+                        SetIsItemSelected(item, false);
+                        DeSelectAllItems(null, item);
+                    }
+                }
+            }
+            else if (treeViewItem != null)
+            {
+                for (int i = 0; i < treeViewItem.Items.Count; i++)
+                {
+                    var item = treeViewItem.ItemContainerGenerator.ContainerFromIndex(i) as TreeViewItem;
+                    if (item != null)
+                    {
+                        SetIsItemSelected(item, false);
+                        DeSelectAllItems(null, item);
+                    }
```

Comment:
could we  easily deduplicate this by using a variable for `treeView` / `treeViewItem`? (loops are duplicated)